Fixes crash when displaying relation editor

troopa81 opened the pull request:
## Description

This PR fixes several crashes related to relation editor when attribute table dialog is called
- From the identify results dialog
- From python iface
- From multi edits on selected feature
- From fill ring map tool

The issue is due of uncomplete QgsAttributeEditorContext, so I add a method in QgisApp to create proper one.

Relation editor is able now to create directly related feature geometry in the canvas, which means that the QgsAttributeDialog must be non modal. So I changed the QgsAttributeDialog flags. I'm not sur it's OK regarding the UI, that why the PR is WIP.

For now, creating geometry directly from relation editor when filling a ring is disable because the code should be modified (remove dialog->exec, create the ring and the new feature at the same time when dialog is closed)

## Checklist

<!-- Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.
-->

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

Commit:
Fixes crash when displaying relation editor

m-kuhn commented on src/gui/qgsattributedialog.cpp:
```diff
@@ -23,7 +23,7 @@
 
 QgsAttributeDialog::QgsAttributeDialog( QgsVectorLayer *vl, QgsFeature *thepFeature, bool featureOwner, QWidget *parent, bool showDialogButtons, const QgsAttributeEditorContext &context )
-  : QDialog( parent )
+  : QDialog( parent, Qt::Window )
```
Out of curiosity, what's that for?

troopa81 replied:
to make it non modal (like in [QgsAttributeTableDialog](https://github.com/troopa81/QGIS/blob/fix_relationeditor_crash_context/src/app/qgsattributetabledialog.h#L56)) so the user can still click on the canvas.

m-kuhn replied:
What about using `Qt::Tool`?

Using this approach will likely make it disappear behind the main window (leaving people with 20 open attribute dialogs - just like they now tend to open the attribute table 20 times)

troopa81 replied:
> Using this approach will likely make it disappear behind the main window 

That's the exact reason of this modification. It's because of this [PR](https://github.com/qgis/QGIS/pull/32528). I want to be able to raise the canvas so user can digitize a feature and then raise back the attribute dialog. 

> (leaving people with 20 open attribute dialogs - just like they now tend to open the attribute table 20 times)

I understand your concern, but I can't see a better way to solve this UI issue. Maybe, we could let the window on top of the canvas but the user can still digitize in the canvas (he has to move the window on the side though) but it's not really user friendly.

m-kuhn replied:
A better approach would be

- hide the window
- set the map tool
- show instructions in the message bar
- when done re-raise the window
- when the map tool is deactivated re-raise the window

troopa81 replied:
Indeed, it seems the best approach, thanks for the hint, I made the modifications

troopa81 commented:
I removed the WIP tag.

> Relation editor is able now to create directly related feature geometry in the canvas, which means that the QgsAttributeDialog must be non modal. So I changed the QgsAttributeDialog flags. I'm not sure it's OK regarding the UI, that why the PR is WIP.

@m-kuhn Regarding this comment, do you think it's OK to make the QgsAttributeDialog non modal? This dialog is called from different place in the application. I made some tests and it looks like OK everytime.

troopa81 commented:
@elpaso I remove the q_assert on cadDockWidget that you add for a previous fix because I enable/disable geometry feature creation from relationeditor according to wether or not caddockwidget has been set. is that OK for you?

m-kuhn commented on src/app/qgisapp.h:
```diff
@@ -1102,6 +1102,12 @@ class APP_EXPORT QgisApp : public QMainWindow, private Ui::MainWindow
     void newBookmark( bool inProject = false );
 
+    /**
+     * Create a default attribute editor context
```
```suggestion
     * Creates a default attribute editor context using the main map canvas and the main edit tools
```

elpaso commented:
@troopa81 ok to remove the assert but we need to do something to prevent a misuse of the API, the checks that you have added are ok, perhaps you should also add a note/warning to the getters documentation to make sure a developer knows that the returned value might be a nullptr and it must be checked before use.

troopa81 commented:
@elpaso I add a warning message on cadDockWidget only, I don't see any reason for the other pointer to be null.

m-kuhn commented on src/app/qgsfeatureaction.cpp:
```diff
@@ -53,22 +53,18 @@ QgsAttributeDialog *QgsFeatureAction::newDialog( bool cloneFeature )
   context.setFormMode( QgsAttributeEditorContext::StandaloneDialog );
 
   QgsAttributeDialog *dialog = new QgsAttributeDialog( mLayer, f, cloneFeature, parentWidget(), true, context );
-  dialog->setWindowFlags( dialog->windowFlags() | Qt::Tool );
```
Is this done somewhere else now?

troopa81 replied:
No, it was from a previous modification, I forget to restore it. It's done now.

m-kuhn commented on src/gui/qgsmaptooladvanceddigitizing.cpp:
```diff
@@ -25,7 +25,6 @@ QgsMapToolAdvancedDigitizing::QgsMapToolAdvancedDigitizing( QgsMapCanvas *canvas
   : QgsMapToolEdit( canvas )
   , mCadDockWidget( cadDockWidget )
 {
-  Q_ASSERT( cadDockWidget );
```
What's the reason for removing this?

troopa81 replied:
I want to disable geometry creation from relation reference widget and relation editor based on the fact that _cadDockWidget_ variable from context is null because of QgsMapToolFillRing (it creates one feature and modify another, it's a mess if you start creating child at this very moment).

After a second thoughts, I change the way I did this to keep the assert. Do you think it's fine this way?

m-kuhn replied:
Yes, that sounds good. I'd just add a docstring (to `setEditorContext`) to make this kind of behavior explicit for other places where we will want the same behavior.

troopa81 replied:
Done! I also add the same kind of behavior for the relation reference widget.

m-kuhn reviewed (approved):
Thank you for the patience @troopa81

troopa81 commented:
@m-kuhn thanks for your review and precious help

mhugo commented:
Merged, since it has been approved. Thanks @troopa81 !
